### PR TITLE
Fix conflict between fp16 and deterministic sampling

### PR DESCRIPTION
**Pull request by Waino**
Fix conflict between fp16 and deterministic sampling

    Due to the removal of the grad hook, MultipleOptimizer no longer has a
    method step, it has been replaced with externally_managed_step which
    takes information about which optimizers need to be stepped. This means
    that it is no longer compatible with torch.cuda.amp.GradScaler.

    While fixing this issue, the MultipleOptimizer system was also
    refactored.
    - MultipleOptimizer and the OpenNMT Optimizer wrapper switched places:
      MultipleOptimizer now wraps the other one, instead of the reverse.
    - The OpenNMT Optimizer was renamed to SubOptimizer for clarity.
    - SubOptimizer handles learning rate scheduling and grad clipping.
    - MultipleOptimizer handles creation of multiple optimizers, grad scaling,
      restoring from checkpoint, backward, zero_grad, deciding which
      suboptimizers to step, and reporting.
    - Each optimizer now individually controls its learning rate schedule.
      When new components with freshly initialized parameters are introduced
      by the curriculum, they now apply warmup to the LR of these
      parameters. This should improve stability.
    - As each optimizer has its own learning rate, it is not obvious what to
      log in the report_training one-liner. Learning rate was removed.
      Instead, all optimizers log their learning rates. This is currently
      log spam, but will be lowered to debug in #70.

    Note that each sub-optimizer having its own GradScaler leads to multiple backward
    passes and RuntimeError. There can only be one GradScaler, which must
    therefore be the responsibility of MultipleOptimizer.

    Closes: #71

### CODE REVIEW

File: mammoth/utils/statistics.py [hunk 1]
```diff
@@ -149,14 +149,14 @@ def output(self, step, num_steps, learning_rate, start, metadata=None):
         if num_steps > 0:
             step_fmt = "%s/%5d" % (step_fmt, num_steps)
         logger.info(
-            ("%s: Step %s; acc: %6.2f; ppl: %5.2f; xent: %4.2f; " + "lr: %7.5f; %3.0f/%3.0f tok/s; %6.0f sec")
+            ("%s: Step %s; acc: %6.2f; ppl: %5.2f; xent: %4.2f; %3.0f/%3.0f tok/s; %6.0f sec")
```
Comment (TimotheeMickus) [inline mammoth/utils/statistics.py:152]: honestly an f-string would be nice ...

File: mammoth/utils/optimizers.py [hunk 2]
```diff
@@ -181,243 +138,255 @@ def linear_warmup_decay(step, warmup_steps, rate, train_steps):
...
+    ):
+        self.suboptimizers = suboptimizers
+        self.grad_scaler = grad_scaler
+        # Global training step is incremented for each minibatch.
+        # There may not be any actual parameters that have been trained this number of steps,
+        # or any optimizer stepped this number of times.
+        self.global_training_step = 1
 
     @classmethod
     def from_opts(cls, model, opts, task_queue_manager, checkpoint=None):
```
Comment (TimotheeMickus) [inline mammoth/utils/optimizers.py:234]: docstring

File: mammoth/utils/optimizers.py [hunk 3]
```diff
@@ -181,243 +138,255 @@ def linear_warmup_decay(step, warmup_steps, rate, train_steps):
...
-        """Returns the current learning rate."""
-        if self._learning_rate_decay_fn is None:
-            return self._learning_rate
-        scale = self._learning_rate_decay_fn(self._decay_step)
-        return scale * self._learning_rate
+        if self.grad_scaler is not None:
+            # Updates the scale for next iteration.
+            self.grad_scaler.update()
+
+    def report_steps(self):
```
Comment (TimotheeMickus) [inline mammoth/utils/optimizers.py:345]: docstring

File: mammoth/utils/optimizers.py [hunk 4]
```diff
@@ -181,243 +138,255 @@ def linear_warmup_decay(step, warmup_steps, rate, train_steps):
...
-            kwargs = {}
-            if "update_master_grads" in fn_args(self._optimizer.backward):
-                kwargs["update_master_grads"] = True
-            self._optimizer.backward(loss, **kwargs)
+        if do_load is True:
+            for k in state_dict["optimizers"].keys():
+                self.suboptimizers[k].load_state_dict(state_dict["optimizers"][k])
         else:
-            loss.backward()
+            logger.info("Some components do not match. Do not load optimizer from checkpoint.")
```
Comment (TimotheeMickus) [inline mammoth/utils/optimizers.py:373]: can we be explicit with the components extra/not found? perhaps in a debug message (to be fixed, ik)

File: mammoth/utils/optimizers.py [hunk 5]
```diff
@@ -1,40 +1,15 @@
...
-            suboptimizers[name] = optimizer
-
-    optimizer = MultipleOptimizer(suboptimizers, None)
-    return optimizer
-
-
-def build_torch_optimizer(model, opts, task_queue_manager):
-    """Builds the PyTorch optimizer.
+def get_base_optimizer(opts):
+    """Builds the PyTorch optimizer factory.
```
Comment (TimotheeMickus) [inline mammoth/utils/optimizers.py:12]: we're building a factory...

**Review by TimotheeMickus — APPROVED** [review]
LGTM.
Would benefit from extra comments here and there, but I don't have much to suggest in terms of impl. i'm a bit wary of `get_base_optimizer` but i don't have better to suggest, tbf.

File: mammoth/utils/optimizers.py (hunk 4 above)
Comment (Waino) [reply mammoth/utils/optimizers.py:373]: I actually didn't add this line, the diff is just wonky.
Seems to be added by @josephattieh in commit e3d8a64f8
